### DISCUSSION TITLE
Using dependencies from npm in user scripts

### DISCUSSION THREAD

**Discussion by pburrows** [chhoumann/quickadd#611] [General]
I have a user script macro that uses [simple-git](https://www.npmjs.com/package/simple-git) to push changes to github. I am struggling on how to reference the simple-git dependency.
I've tried doing an `npm i simple-git` in the same directory as the script file, in the parent directory, in the `.obsidian/modules` directory. I've even tried copying the index.js file from simple-git into the same folder as my user script and doing `const simpleGit = require('./simple-git')`

Nothing seems to work. The error is always "cannot find module 'simple-git'"

```bash
node:internal/modules/cjs/loader:1082 Uncaught (in promise) Error: Cannot find module 'simple-git'
Require stack:
- electron/js2c/renderer_init
```

Any tricks on how to reference dependencies in user script macros?

**Comment by chhoumann** [comment]
Coming back to this one very late - sorry about that.

The reason nothing you tried worked: your script isn't loaded by Node's module system, so `require("simple-git")` (and even `require("./simple-git")`) resolves against Electron's internals - that's the `electron/js2c/renderer_init` in your stack trace - not against your script's folder. A vault-local `node_modules` is never searched.

Two patterns that do work:

**Option A - require by absolute path.** Run `npm i simple-git` in a folder inside your vault (say `scripts/`), then build the absolute path at runtime:

```js
module.exports = async ({ app }) => {
  const path = require("path");
  const simpleGit = require(
    path.join(app.vault.adapter.getBasePath(), "scripts", "node_modules", "simple-git")
  );
  const git = simpleGit(app.vault.adapter.getBasePath());
  // ... git.add("./*").commit("...").push() etc.
};
```

**Option B - bundle.** Build your script into a single file and point QuickAdd at the bundle, so no `node_modules` lives in the vault at all:

```
esbuild myscript.js --bundle --platform=node --format=cjs --outfile=myscript.bundle.js
```

Either way this is desktop-only: `require` (and simple-git shelling out to git) doesn't exist on mobile, so guard with `Platform.isDesktopApp` if the vault is shared with a phone.

The require-by-absolute-path pattern and its caveats are documented here: https://quickadd.obsidian.guide/docs/UserScripts
